zaps/0002-encrypt-push-notifications.md

```diff
@@ -0,0 +1,263 @@
+- ZAP PR: [zulip/zulip-architecture#2](https://github.com/zulip/zulip-architecture/pull/2)
+
+# Summary
+
+Proposal to add end-to-end encryption for mobile push notifications.
+
+# Motivation
+
+Currently, the Zulip Push Notification Forwarding Service receives
+(but does not store or log) the contents of individual mobile push notifications.
+
+Adding end-to-end encryption for push notifications will ensure that the
+actual message content does not need to be readable by the push notification
+bouncer (or Google/Apple's servers, for that matter).
+
+# Detailed design
+
+## High-Level Plan
+
+- Client checks the server if it supports E2E encryption. If it doesn't,
+  use the legacy push notification registration flow.
+
+- Client generates an asymmetric key pair (public and private key).
+  It stores the private key securely and makes an API call to the server
+  with the public key to register itself for push notifications.
+
+- While sending notification, the server encrypts the notification content
+  to the public key from above and sends it to push bouncer.
+  Push bouncer forwards it to FCM/APNs.
+
+- Client receives the notification via FCM/APNs and decrypts the content
+  using the private key it stored earlier.
+
+## Data model
+
+- On the client, include the following fields on [the Account type](https://github.com/zulip/zulip-flutter/blob/ec9aa351b3050a0fa3e467007826576110fdbf99/lib/model/database.dart#L66-L94):
+  ```
+  model Account:
+    …
+    push_account_id: int  # 64-bit random integer generated by client; unique here.
+    last_acked_push_token: str  # The last push token the server acknowledged receipt of from this client.
+    push_private_key: PrivateKey  # Private part of the asymmetric key pair generated by the client.
+  ```
+- A helper data structure
+  ```
+  struct PushRegistration:
+    token_kind: "apns" | "fcm"
+    token: str
+    timestamp: datetime
+  ```
+
+- The push bouncer has an asymmetric key pair, of which the
+  public part (`bouncer_public_key`) is baked into clients.
+
+- On the server, add a new table (coexists with `PushDeviceToken`):
+  ```
+  class zerver.models.push_notifications.PushDevice:
+    user: ForeignKey[UserProfile]
+    encrypted_push_registration: Ciphertext  # A PushRegistration, encrypted with bouncer_public_key.
```

Comment:
Here's an adjustment that wasn't in the original draft:
```suggestion
    encrypted_push_registration: Ciphertext | None # A PushRegistration, encrypted with bouncer_public_key. Null once bouncer has confirmed the registration.
```

This field is never gets used once `bouncer_device_id` is set to non-null. So it can be cleared out at that point.

More conceptually: this field basically describes a pending request that needs to be sent to the bouncer. Once that request has been successfully made, and the acknowledgement recorded in the form of `bouncer_device_id`, that information is obsolete.